### PR TITLE
Make the s_server command listen on IPv6 only when requested

### PR DESCRIPTION
**Pull request by bernd-edlinger**
This makes the server listen on IPv4 and IPv6 when neither -4 nor -6 is used.
When -6 is used the server will listen on IPv6 only.

Depends on #5139 for proper operation on Windows.

### CODE REVIEW

File: apps/s_socket.c [hunk 1]
```diff
@@ -175,10 +179,27 @@ int do_server(int *accept_sock, const char *host, const char *port,
...
 
-    asock = BIO_socket(BIO_ADDRINFO_family(res), BIO_ADDRINFO_socktype(res),
-                       BIO_ADDRINFO_protocol(res), 0);
+    sock_family = BIO_ADDRINFO_family(res);
+    sock_type = BIO_ADDRINFO_socktype(res);
+    sock_protocol = BIO_ADDRINFO_protocol(res);
+    sock_address = BIO_ADDRINFO_address(res);
+    next = BIO_ADDRINFO_next(res);
+    if(sock_family == AF_INET6)
+        sock_options |= BIO_SOCK_V6_ONLY;
```
Comment (dot-asm) [inline apps/s_socket.c]: I don't quote follow. My understanding is that `if (family == AF_INET6) sock_options |= BIO_SOCK_V6_ONLY;` would be sufficient. I mean when you do lookup, you ask for AF_INET6 and it's going to fail if there is none. And if you ask for *any* address and got AF_INET6, then you shouldn't set BIO_SOCK_V6_ONLY anyway... No?

Comment (bernd-edlinger) [reply apps/s_socket.c]: I see this needs still a comment...
The problem here is:
There are cases when only AF_INET6 and nothing else is returned. That gets
BIO_SOCK_V6_ONLY.
But there are cases where AF_INET and AF_INET6 are returned (Linux),
and there are cases where AF_INET6 and AF_INET are returned (Windows).
Both should use the ANY INET6 address, and no BIO_SOCK_V6_ONLY.
Now understand?

Comment (levitte) [reply apps/s_socket.c]: I think I agree with @dot-asm.  The address that `sock_address` presents will determine the stack to use anyway...

Comment (bernd-edlinger) [reply apps/s_socket.c]: Currently when no -4 or -6 is used, we do only look at the first address that is
returned, so linux listens on IPv4 only, and Windows listens on IPv6 only,
the other PR makes Windows listen on IPv4 and IPv6 (because BIO_SOCK_V6_ONLY
is not used)
The idea is that -6 should listen to IPv6 only and without -4 or -6 per default
we listen to IPv4 and IPv6, independent of the underlying O/S.

Comment (bernd-edlinger) [reply apps/s_socket.c]: > I mean when you do lookup, you ask for AF_INET6 and it's going to fail if there is none.

The interesting case is when AF_UNSPEC is requested.

Comment (dot-asm) [reply apps/s_socket.c]: > Currently when no -4 or -6 is used, we do only look at the first address that is
returned, so linux listens on IPv4 only, and Windows listens on IPv6 only,

Well, if you didn't specify -4 or -6,  it's not unnatural to assume that you might get different results on different systems.

> The idea is that -6 should listen to IPv6 only and without -4 or -6 per default
we listen to IPv4 and IPv6, independent of the underlying O/S.

In other words it's like skewing lookup toward IPv6 and not passing BIO_SOCK_V6_ONLY when you didn't ask for it. So it still sound like `if (family == AF_INET6) sock_options |= BIO_SOCK_V6_ONLY;` would be sufficient... I mean you have to explicitly ask for AF_INET6 *and* resolve it anyway...

As for "skewing", we are really talking about wild-char bindings, right? In other words question is if user is entitled to expect that wild-char would accept both IPv4 and IPv6 connections. But then question also is who is the user? One that uses apps/openssl or one who creates BIO socket? What I'm hinting at is that *maybe* it's appropriate to move that kind of heuristics to b_sock2.c? Maybe not, but let's think about it...

Comment (bernd-edlinger) [reply apps/s_socket.c]: This affects the stability of the test suite:

For instance if the Perl IO:Sockets:IP module is installed,
and you change /eth/hosts
to conatain the line
`::1  localhost`
Then `make test` fails, because it depends on what the name resolution
exactly does with "localhost" and the freedom the s_server command has
per default.

**Comment by levitte** [comment]
From all the discussion, it seems to me like the better action is to create an array of sockets, one for every address that's been returned by `BIO_lookup_ex`, `listen()` on them all, and then `accept()` on the one that gets the connection.  `select()` on the accept sockets would be the way to go for this...

I've had some thoughts on making a PR that does that, but haven't come around to it yet.  I've already enough of a back log, so if anyone else is willing to try that idea, please do.

File: apps/s_socket.c (hunk 1 above)
Comment (bernd-edlinger) [reply apps/s_socket.c]: @dot-asm if I explicitly ask for AF_INET6 (per -6) everything is simple.
I listen on ANY_IPV6 with BIO_SOCK_V6_ONLY.
When I explicitly ask for AF_INET (per -4) everything is even more simple.
But when I do not want to specify the protocol I ask for AF_UNSPEC,
and dependent on the O/S it may either return AF_INET or AF_INET6
or both.  But Windows and Linux return a opposite ordered list.
By this logic I get consistent behavior when both protocols are available.

**Comment by bernd-edlinger** [comment]
Okay @levitte this is only a simple solution for a not really simple problem,
but I think it would still be an improvement over the current state.

For the client side it really becomes much more difficult, and probably there
is no other way than to try to connect to all possible protocols in parallel.
There I see the need for another PR eventually.

**Comment by dot-asm** [comment]
> it seems to me like the better action is to create an array of sockets,

Not to me. I would consider this as over-doing it. It's really about wild-char case and all [important] OSes are perfectly capable of creating socket that accepts both IPv4 and IPv6 and it doesn't seem impossible to achieve consistent behaviour, consistent as it appears to user.

**Comment by richsalz** [comment]
So let me make sure I understand.  If you specify -4 or -6 the server *only* listens on that address.  If you specify neither flag, what it does is system-dependant.  Do I have that right?  If so, +1 from me.

**Comment by bernd-edlinger** [comment]
Yes, also -6 is currently system dependent, 
windows listens on IPv6 only (that broke my cygwin test)
linux listens on IPv4 and IPv6, (which is also not what one would expect from -6)

File: apps/s_socket.c (hunk 1 above)
Comment (dot-asm) [reply apps/s_socket.c]: > But Windows and Linux return a opposite ordered list.

Right.

> By this logic I get consistent behavior when both protocols are available.

By logic below the line in question. I mean you need to see if there is AF_INET6 somewhere, and if there is, you go for for AF_INET6 socket but *without* BIO_SOCK_V6_ONLY, right? In other words BIO_SOCK_V6_ONLY needs to be set only when you explicitly ask for AF_INET6 [with `family`]. No?

Comment (dot-asm) [reply apps/s_socket.c]: What I'm saying is that logic appears over-complicated... And counter-intuitive...

Comment (bernd-edlinger) [reply apps/s_socket.c]: If there is AF_INET in the first and AF_INET6 in the second element (linux)
I use the INET6 address AND do not set V6_ONLY because AF_INET is also advertised.
If there is AF_INET6 in the first and AF_INET in the second element (windows)
I clear the V6_ONLY and use the AF_INET6 address.
If it is more complicated than that, I give up, and just use the first address as before.

**Comment by levitte** [comment]
> For the client side it really becomes much more difficult, and probably there is no other way than to try to connect to all possible protocols in parallel.

Not needed, `s_client` already tries one address after another until a connection happens.  Telnet-style

File: apps/s_socket.c (hunk 1 above)
Comment (dot-asm) [reply apps/s_socket.c]: > logic appears over-complicated... And counter-intuitive...

Consider this. You did *not* pass -6, and lookup returned AF_INET6 alone. Is it appropriate set BIO_SOCK_V6_ONLY? After all, you did *not* pass -6... What if IPv4 interface is added afterwards? ;-)

Comment (bernd-edlinger) [reply apps/s_socket.c]: > What I'm saying is that logic appears over-complicated... And counter-intuitive...

I am interested in your suggestions.

Comment (bernd-edlinger) [reply apps/s_socket.c]: I don't think the result depends on the available interfaces.
I see from the strace just probing if a socket of that family can be created,
this may change if ipv4 is a loadable kernel module, but I would not care.

Comment (dot-asm) [reply apps/s_socket.c]:
```
    if (family == AF_INET6) {
        sock_options |= BIO_SOCK_V6_ONLY;
    } else if (family == AF_UNSPEC && BIO_ADDRINFO_family(res) == AF_INET) {
        /*
         * See if there is even AF_INET6 hit, and if there is, prefer it
         * over AF_INET so that socket listens on both...
         */
        ...
    }
```
Naturally provided that #5139 is in place...

**Comment by richsalz** [comment]
When you say "is current system dependant" do you mean even after this PR?

**Comment by bernd-edlinger** [comment]
No, I meant before this PR, after this pr and #5139, the behavior will be as follows
- when no preference is given, and both available, listen on IPv4 and IPv6
- when no preference and only IPv4 available listen on IPv4 only
- when no preference and only IPv6 available listen on IPv6 only
- when -6 is given and available, listen on IPv6 only
- when -4 is given and available listen on IPv4 only
- otherwise fail

File: apps/s_socket.c (hunk 1 above)
Comment (bernd-edlinger) [reply apps/s_socket.c]: Hmmmmmm.... it may be possible to save one if by using the requested socket family,
but in principal the information content in the result list should be the same.
This might make an assumption that the result of BIO_lookup_ex depends only on the
requested socket family.

**Comment by bernd-edlinger** [comment]
Merged to master and 1.1.0
Thanks!

File: apps/s_socket.c [hunk 2]
```diff
@@ -175,10 +179,27 @@ int do_server(int *accept_sock, const char *host, const char *port,
...
+    if (next != NULL
+        && BIO_ADDRINFO_socktype(next) == sock_type
+        && BIO_ADDRINFO_protocol(next) == sock_protocol) {
+        if (sock_family == AF_INET && BIO_ADDRINFO_family(next) == AF_INET6) {
+            sock_family = AF_INET6;
+            sock_address = BIO_ADDRINFO_address(next);
+        }
+        else if (sock_family == AF_INET6 && BIO_ADDRINFO_family(next) == AF_INET)
+            sock_options &= ~BIO_SOCK_V6_ONLY;
```
Comment (dot-asm) [inline apps/s_socket.c]: Three style nits here.

**Comment by kroeckx** [comment]
I have to agree with @levitte that you should create multiple sockets, and set the IPv6 socket to only do IPv6. This will not work on some OSs, where an IPv6 socket really only does IPv6.

I also planned to make BIO_new_accept() do this, but never got around to doing it.

**Comment by bernd-edlinger** [comment]
@kroeckx this is not set in stone, I just did it to fix certain corner cases in the TLS test cases.

File: apps/s_socket.c (hunk 2 above)
Comment (bernd-edlinger) [reply apps/s_socket.c]: Yes, exactly.  Please see #5173
